apm-agent-core/src/main/java/co/elastic/apm/agent/bci/ElasticApmAgent.java

```diff
@@ -257,8 +259,19 @@ private static synchronized void initInstrumentation(final ElasticApmTracer trac
         Runtime.getRuntime().addShutdownHook(new Thread(ThreadUtils.addElasticApmThreadPrefix("init-instrumentation-shutdown-hook")) {
             @Override
             public void run() {
+                if (tracer.getConfig(CoreConfiguration.class).logUsedInstrumentationGroupsOnExit()) {
+                    Set<String> usedInstrumentationGroups = new TreeSet<>();
+                    for (ElasticApmInstrumentation instrumentation : instrumentations) {
+                        if (!instrumentation.isUsed()) {
+                            continue;
+                        }
+                        usedInstrumentationGroups.addAll(instrumentation.getInstrumentationGroupNames());
+                    }
+                    logger.info("Used instrumentation groups: {}", usedInstrumentationGroups);
```

Comment:
This set doesn't contain the minimum set of instrumentation groups that are enabled. The reason is that often an instrumentation has a generic group, such as `http-client`, and a more specific one, such as `okhttp`. If only `okhttp` had a match, but `apache-httpclient` did not, we would not want to include `http-client`, as it would enable both, even though `apache-httpclient` never matched.

Therefore, after compiling the list of used instrumentation groups, remove all the groups that did not match.
For example `[http-client, okhttp] - [http-client, apache-httpclient] = [okhttp]`.

Reply:
This sounds very useful.

What if `CaptureSpanInstrumentation` is used but `CaptureTransactionInstrumentation` is not? Both have "public-api" and "annotations" as instrumentation groups. What should be logged in this case? Both groups?

Reply:
Excellent question. Yes, both groups should be logged.

You could implement this by maintaining two `Set<Collection<String>>`, one for matched and one for non-matched instrumentation groups. As both `CaptureSpanInstrumentation` and `CaptureTransactionInstrumentation` have the same collection of group names, they're collapsed into one entry of `["public-api", "annotations"]`. So when one instrumentation matches, the instrumentation groups are moved out of the set of non-matched instrumentation groups and get added to the set of matched groups.

Reply:
Done